src/pf/existing/detect_base_dir.cpp

```diff
@@ -0,0 +1,53 @@
+#include "./detect_base_dir.hpp"
+
+#include <algorithm>
+#include <cstdlib>  // std::abort
+#include <fstream>
+#include <string>
+#include <utility>
+
+#include <boost/range/iterator_range.hpp>
+
+namespace fs = pf::fs;
+
+namespace {
+std::optional<fs::path> parse_cmakecache_homedir(fs::path const& cmakecache) {
+    std::string line;
+
+    std::ifstream file{cmakecache};
+
+    while (std::getline(file, line)) {
+        // if (line.starts_with("CMAKE_HOME_DIRECTORY"))
+        if (line.rfind("CMAKE_HOME_DIRECTORY", 0) == 0) {
+            auto index = line.find('=');
+            line.erase(line.begin(), line.begin() + index + 1);
+            return fs::path{line};
+        }
+    }
+
+    return std::nullopt;
+}
+}  // namespace
+
+std::optional<fs::path> pf::detect_base_dir(fs::path from_dir) {
+    for (auto const& cur_dir :
+         boost::make_iterator_range(pf::ascending_iterator{from_dir}, pf::ascending_iterator{})) {
```

Comment:
I think it may be simpler to just recurse (checking each parent directory) do find the root `CMakeLists.txt` rather than try and walk it with a special iterator.